src/main/java/picard/sam/markduplicates/CheckDuplicateMarking.java

```diff
@@ -0,0 +1,173 @@
+    static final String USAGE_SUMMARY = "Checks the consistency of duplicate markings.";
+    static final String USAGE_DETAILS = "This tool checks that all reads with the same queryname have their duplicate marking flags set the same way. " +
+            "NOTE: This tool does NOT check that the duplicate marking is correct. The ONLY thing that it checks is that the 0x400 bit-flags of records " +
+            "with the same queryname are equal.";
+
+    @Argument(doc = "Input BAM or SAM file to check.", shortName = StandardOptionDefinitions.INPUT_SHORT_NAME)
+    public File INPUT;
+
+    @Argument(doc = "Output file into which bad querynames will be placed (if not null).", shortName = StandardOptionDefinitions.OUTPUT_SHORT_NAME, optional = true)
+    public File OUTPUT = null;
+
+    @Argument(doc = "Which reads of the same name should be checked to have same duplicate marking.")
+    public Mode MODE = Mode.ALL;
+
+    private String currentReadName = "";
+    private boolean currentReadDuplicateMarked = false;
+
+    private static final Log log = Log.getInstance(CheckDuplicateMarking.class);
+    private final ProgressLogger progress = new ProgressLogger(log, (int) 1e6, "Checked.");
+
+    private static final int NUM_WARNINGS = 100;
+    private int numBadRecords = 0;
+
+    public enum Mode implements CommandLineParser.ClpEnum {
+        ALL("Check all reads."),
+        PRIMARY_ONLY("Check primary alignments."),
+        PRIMARY_MAPPED_ONLY("Check mapped alignments."),
+        PRIMARY_PROPER_PAIR_ONLY("Check mapped alignments.");
+
+        private final String message;
+
+        Mode(final String message) {
+            this.message = message;
+        }
+
+        public String getHelpDoc() {
+            return message;
+        }
+    }
+
+    @Override
+    protected int doWork() {
+        IOUtil.assertFileIsReadable(INPUT);
+        if (OUTPUT != null) {
+            IOUtil.assertFileIsWritable(OUTPUT);
+        }
+
+        try (SamReader reader = SamReaderFactory.makeDefault().referenceSequence(REFERENCE_SEQUENCE).open(INPUT)) {
+
+            checkDuplicateMarkingsInIterable(getSortedRecordsFromReader(reader));
+            if (numBadRecords > 0) {
+                log.error("Found " + numBadRecords + " records that do not agree on their duplicate flag.");
+            } else {
+                log.info("All records' duplicate markings agree.");
+            }
+
+        } catch (IOException e) {
+            throw new PicardException("Error while reading input file " + INPUT, e);
+        }
+
+        return numBadRecords > 0 ? 1 : 0;
+    }
+
+    private Iterable<SAMRecord> getSortedRecordsFromReader(final SamReader reader) {
+        if (reader.getFileHeader().getSortOrder() == SAMFileHeader.SortOrder.queryname) {
+            return reader;
+        }
+
+        log.info("Input file isn't queryname sorted. Sorting into temp space.");
+
+        final Path[] tmpPaths = TMP_DIR.stream()
```

Comment:
If this is done in more than one CLP, it would make sense to move it to `CommandLineProgram`.

(Also, it is odd that `File` is deprecated in `SortingCollection` but not in `SortingLongCollection`)

Reply:
it's done in only two other CLPs....somehow.

Reply:
hmmm. most CLPs use system property "java.io.tmpdir" but I chose to use input variable TMP_DIR...not sure which is right.

Reply:
oh...it's the same. I'll use the other.